README.md
Folder contains:
cd.c -> codes for cd command
echo.c -> codes for echo command
hist.c -> codes for history command
ls.c -> codes for ls command
pinfo.c -> codes for pinfo command (user defined)
pwd.c ->codes for pwd command
history.txt -> store last (at max) 20 commands
shell.c -> code for basic structure of shell
oth.c -> codes for all other commands(using execvp)

TO close shell type->  exit

HOW TO RUN SHELL:
-->Run make command,(Compilation)
-->A executable file named shell is made,
-->Run ./shell
-->Shell STARTED

src/history.txt
emacs
gcc shell.c
./a.out
ls
exit
pinfo
exit
gcc shell.c
./a.out
exit
exit
gcc shell.c
pinfo
pinfo 28729
./a.out
pinfo
pinfo 28736
exit
pinfo
exit


src/ls.c
void list(char *str,int ptr)
{
	int flag_l=0,flag_a=0;
	int temp=ptr;
	for (int i = ptr; i < strlen(str); ++i)
	{
		if(str[i]=='-')
		{
			if(str[i+1]=='a' && str[i+2]=='l')
			{
				flag_a=1;
				flag_l=1;
			}
			else if(str[i+1]=='l' && str[i+2]=='a')
			{
				flag_a=1;
				flag_l=1;
			}
			else if(str[i+1]=='a')
			{
				flag_a=1;
			}
			else if(str[i+1]=='l')
			{
				flag_l=1;
			}
		}
	}
	if(flag_a==0 && flag_l==0)
	{
		int count=0,x;
		for (int i = ptr; i < strlen(str); ++i)
		{
			if(str[i]!=' ')
			{
				while(str[i]!=' ')
				{
					i++;
				}
				count++;
			}
		}
		int g=0;
		x=count;
		struct dirent *de;
		char path[1024],real_path[1024];
		char* token;
		char* qwerty=str;
		if(count==0)
			{
				DIR *dr = opendir(".");
				while ((de = readdir(dr)) != NULL)
			  	{
			  		if(de->d_name[0]!='.')
						printf("%s  ", de->d_name); 
			  	}
				printf("\n");
			}
		else
		{
			while((token=strtok_r(qwerty," ",&qwerty)))
			{
				g++;
				if(g==1)
					continue;
				if(token[0]=='~')
				{
					for (int i = 0; i < strlen(actualpath); ++i)
					{
						real_path[i]=actualpath[i];
					}
					real_path[strlen(actualpath)]='\0';
					if(strlen(token)!=1)
					for (int i = 1; i < strlen(token); ++i)
					{
						real_path[strlen(actualpath)-1+i]=token[i];
						if(i==strlen(token)-1)
							real_path[strlen(actualpath)+i]='\0';
					}
				}
				else
				{
					char *ptr1;
					char *symlinkpath = token;
					ptr1 = realpath(symlinkpath, real_path);
			    }
			    DIR *dr = opendir(real_path);
			    if (dr == NULL)  
			    {
			    	if(access( real_path, F_OK ) != -1 )
			        	{
			        		printf("%s\n",token);
			        	}
			        else
			        	{
			        		printf("No such file or directory existss\n");
			        	}
			    	continue;
			    }
			    if(count!=1)
			    	printf("%s:\n",token);
			  	while ((de = readdir(dr)) != NULL) 
			  	{
			  		if(de->d_name[0]!='.')
						printf("%s  ", de->d_name); 
			  	}
			  	printf("\n");
				closedir(dr);
			    if(count!=1)
			    {
			    	if(x>1)
			    	{
			    		x--;
			    		printf("\n");
			    	}
			    }
			}
		}
	}
	else if(flag_l==0 && flag_a==1)
	{
		struct dirent *de;
		DIR *dr = opendir(".");
		while ((de = readdir(dr)) != NULL) 
	  	{
			printf("%s  ", de->d_name); 
	  	}
		printf("\n");
	}
	else if(flag_l==1)
	{
		char date_str[1024];
		struct dirent *de;
		struct stat sfile;
		DIR *dr = opendir(".");
		while ((de = readdir(dr)) != NULL) 
	  	{
	  		if(de->d_name[0]=='.' && flag_a==0)
	  			continue;
			char temp_filename[1024];
			strcpy(temp_filename,de->d_name);
			if(stat(temp_filename,&sfile)==-1)
			{
				printf("ERROR: file is not accessible\n");
			}
			printf((sfile.st_mode & S_IFDIR)? "d":"-");
		  	printf((sfile.st_mode & S_IRUSR)? "r":"-");
		  	printf((sfile.st_mode & S_IWUSR)? "w":"-");
		  	printf((sfile.st_mode & S_IXUSR)? "x":"-");
		  	printf((sfile.st_mode & S_IRGRP)? "r":"-");
		  	printf((sfile.st_mode & S_IWGRP)? "w":"-");
		  	printf((sfile.st_mode & S_IXGRP)? "x":"-");
		  	printf((sfile.st_mode & S_IROTH)? "r":"-");
		  	printf((sfile.st_mode & S_IWOTH)? "w":"-");
		  	printf((sfile.st_mode & S_IXOTH)? "x":"-");
		  	printf(" ");
		  	printf("%u",(unsigned int)sfile.st_nlink);
		  	printf(" ");
		  	struct passwd *pw = getpwuid(sfile.st_uid);
			struct group *gr = getgrgid(sfile.st_gid);
			printf("%s %s ",pw->pw_name,gr->gr_name);
			printf("%5ld ",sfile.st_size );
			strcpy(date_str,ctime(&sfile.st_ctime));
			date_str[7]='\0';
			date_str[10]='\0';
			date_str[16]='\0';
			printf("%s ",date_str +4);
			printf("%s ",date_str +8 );
			printf("%s ",date_str +11 );
			printf("%s\n",temp_filename);
	  	}
	}
	return;
}

src/echo.c
void f_echo(char *str1,int ptr,int l)
{
	int q=ptr,x,flag=0,flag_x=-1;
	for (int i = ptr-1; i < strlen(str1); ++i)
	{
		if(str1[i]=='\'')
		{
			flag_x=0;
			i++;
			while(i<strlen(str1) && str1[i]!='\'')
			{
				i++;
			}
			if(i<strlen(str1) && str1[i]=='\'')
			{
				flag_x=1;
			}
		}
		else if(str1[i]=='\"')
		{
			flag_x=0;
			i++;
			while(i<strlen(str1) && str1[i]!='\"')
			{
				i++;
			}
			if(i<strlen(str1) && str1[i]=='\"')
			{
				flag_x=1;
			}
		}
	}
	if(flag_x==0)
	{
		printf("Invalid Input\n");
		return;
	}
	for (int i = q; i < l; ++i)
	{
		flag=0;
		while(i<strlen(str1) && str1[i]==' ')
			{
				flag=1;
				i++;
			}
		if(flag==1)
			printf(" ");
		if(str1[i]=='\'')
		{
			x=i+1;
			i++;
			while(i<strlen(str1) && str1[i]!='\'')
			{
				i++;
			}
			if(i<strlen(str1) && str1[i]=='\'')
			{
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]='\'';
			}
			else
				return;
		}
		else if(str1[i]=='\"')
		{
			x=i+1;
			i++;
			while(i<strlen(str1) && str1[i]!='\"')
			{
				i++;
			}
			str1[i]='\0';
			printf("%s",str1+x);
			str1[i]='\"';
		}
		else
		{
			x=i;
			while(i<strlen(str1) && str1[i]!='\"' && str1[i]!='\'' && str1[i]!=' ')
			{
				i++;
			}
			if(str1[i]==' ')
			{
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]=' ';
				i--;
			}
			else if(str1[i]=='\'')
			{
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]='\'';
				i++;
				x=i;
				while(i<strlen(str1) && str1[i]!='\'')
				{
					i++;
				}
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]='\'';
			}
			else if(str1[i]=='\"')
			{
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]='\"';
				i++;
				x=i;
				while(i<strlen(str1) && str1[i]!='\"')
				{
					i++;
				}
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]='\"';
			}
			else if(str1[i]=='\0')
			{
				str1[i]='\0';
				printf("%s",str1+x);
				str1[i]=' ';
				i--;
			}
		}
	}
	printf("\n");	
}

src/hist.c
void hist(char *str1,int ptr,int flag_h)
{
		char str[1024],buffer[1024],count=0,temp[100],path[1024],path2[1024];
		strcpy(str,str1);
		int l=strlen(str);
		for (int i = 0; i < strlen(actualpath); ++i)
		{
			path[i]=actualpath[i];
			path2[i]=actualpath[i];
		}
		strcpy(temp,"/history.txt");
		for (int i = 0; i <= strlen(temp); ++i)
		{
			path[i+strlen(actualpath)]=temp[i];
			path2[i+strlen(actualpath)]=temp[i];
		}
		path2[strlen(actualpath)+strlen(temp)-1]='x';
	if(flag_h==0)
	{
		FILE * fp;
		fp = fopen (path,"ab+");
		char c[1024];
		while(fgets(buffer,1020,fp)!=NULL)
		{
			count++;
		}
		if(count<20)
		{
			fprintf(fp, "%s\n",str );
			count++;
			fclose(fp);
		}
		else 
		{
			FILE * fp2;
			fp2 = fopen (path2,"ab+");
			int g=0;
			fseek(fp2,0,0);
			fseek(fp,0,0);
			while(fgets(buffer,1020,fp)!=NULL)
			{
				++g;
				if(g==1)
				continue;
				fprintf(fp2, "%s", buffer);
			}
			fprintf(fp2, "%s\n",str );
			fclose(fp);
			fclose(fp2);
			remove(path);
			rename(path2,path);
		}
	}
	else if(flag_h==1)
	{
		int val=-1,lines=0,limit=0,flag_n=0;
		if(ptr==strlen(str1))
			{
				flag_n=1;
			}
		else
			val=atoi(str+ptr);
		if(val<=0 && flag_n!=1 && str[ptr]!='0')
			{
				printf("INVALID INPUT\n");
				return;
			}
		FILE * fp;
		fp=fopen(path,"ab+");
		while(fgets(buffer,1020,fp)!=NULL)
		{
			lines++;
		}
		fseek(fp,0,0);
		if(flag_n==1)
		{
			if(lines>=10)
				val=10;
			else
				val=lines;
		}
		if(lines>=lines-val)
		{
			if(val>10)
			{
				val=10;
				printf("Printing last 10 commands\n");
			}
			limit=0;
			while(fgets(buffer,1020,fp)!=NULL)
			{
				limit++;
				if(limit>lines-val)
					printf("%s",buffer);
			}
		}
		else if(lines<val)
		{
			printf("Only %d commands are saved\n",lines );
		}
		fclose(fp);

	}
	return;
}

src/pinfo.c
void pinfo(char *str,int ptr)
{
	int l=strlen(str);
	char str1[1024];
	if(ptr==l)
	{
		int pid=getpid();
		printf("pid -- %d\n",getpid());
		sprintf(str1,"/proc/%d",getpid());
	}
	else
	{
		strcpy(str1,"/proc/");
		printf("pid -- ");
		for (int i = ptr; i < strlen(str); ++i)
		{
			printf("%c",str[i]);
			str1[6+i-ptr]=str[i];
			if(i==l-1)
				str1[7+i-ptr]='\0';
		}
		printf("\n");
	}
	int count=0,ct=0,flag=0;
	char str3[1024];
	strcpy(str3,str1);
	l=strlen(str3);
	str3[l]='/';
	str3[l+1]='s';
	str3[l+2]='t';
	str3[l+3]='a';
	str3[l+4]='t';
	str3[l+5]='u';
	str3[l+6]='s';
	str3[l+7]='\0';
	FILE *fp = fopen(str3,"r");
    char word[10000];
    char str2[1024];
        while( fscanf(fp, "%s", word) != EOF )
        {
        	count++;
        	if(!strcmp("State:",word))
        	{
        		ct++;
        		flag=1;
        		continue;
        	}
        	else if(!strcmp("VmSize:",word))
        	{
        		ct++;
        		flag=2;
        		continue;
        	}
        	if(flag!=0)
        	{
        		if(flag==1)
        		{
        			printf("Process Status -- %s\n",word );	
        			flag=0;
        		}
        		else if(flag==2)
        		{
        			printf("memory -- %s  {virtual memory}\n", word);
        			flag=0;
        		}
        	}
        }
        if(ct!=2)
        {
        	printf("memory -- ?\n");
        }
        l=strlen(str1);
        str1[l]='/';
        str1[l+1]='e';
        str1[l+2]='x';
        str1[l+3]='e';
        str1[l+4]='\0';
        memset(str2, 0, sizeof(str2));
		int x= readlink(str1,str2,1023);
		flag=1;
		long long int n1,n2,i;
		char net_cwd[1024];
		n1=strlen(str2);
		n2=strlen(actualpath);
		for (i = 0; i < n2 && i < n1; ++i)
		{
			if(str2[i]!=actualpath[i])
				{
					flag=0;
					break;
				}
		}
		if(i==n2 && flag==1)
		{
			net_cwd[0]='~';
			for (int j = n2; j < n1+1; ++j)
			{
				net_cwd[j-n2+1]=str2[j];
			}
		}
		else
		{
			for (int j = 0; j < n1+1; ++j)
			{
				net_cwd[j]=str2[j];
			}
		}
		if(x<0)
			perror("Error");
		else
		printf("Executable Path -- %s\n",net_cwd );
}

src/shell.c
#include<stdlib.h>
#include<stdio.h> 
#include<fcntl.h> 
#include<sys/types.h>
#include <sys/wait.h> 
#include<sys/stat.h>
#include<unistd.h>
#include<string.h>
#include<libgen.h>
#include<pwd.h>
#include<dirent.h> 
#include<grp.h>
#include <time.h>
char actualpath[1024];
#include "cd.c"
#include "pwd.c"
#include "echo.c"
#include "ls.c"
#include "pinfo.c"
#include "hist.c"
long long int T=0;
void commandprompt(char *arr)
{
	if(T==0)
		{
			char *ptr1;
			char *symlinkpath = arr;
			ptr1 = realpath(symlinkpath, actualpath);
			int point;
			for (int i = strlen(actualpath)-1; i >=0; --i)
			{
				if(actualpath[i]=='/')
				{
					actualpath[i]='\0';
					point=i;
					break;
				}
			}
		}
	T++;
	char hostname[1024], *username,cwd[1024],net_cwd[1024];
	username=getlogin();
	gethostname(hostname, 1023);
	long long int n1,n2,flag=1,i;
	getcwd(cwd,sizeof(cwd));
	n1=strlen(cwd);
	n2=strlen(actualpath);
	for (i = 0; i < n2 && i < n1; ++i)
	{
		if(cwd[i]!=actualpath[i])
			{
				flag=0;
				break;
			}
	}
	if(i==n2 && flag==1)
	{
		net_cwd[0]='~';
		for (int j = n2; j < n1+1; ++j)
		{
			net_cwd[j-n2+1]=cwd[j];
		}
	}
	else
	{
		for (int j = 0; j < n1+1; ++j)
		{
			net_cwd[j]=cwd[j];
		}
	}
	// fflush(stdout);
	printf("<%s@%s:%s> ",username, hostname,net_cwd);
	fflush(stdout);
}
void handler(int sig)
{
	pid_t pid;
	int status;
	char *exit=(char *)malloc(1024*(sizeof(char)));
	char *exit_status=(char *)malloc(1024*(sizeof(char)));
	pid=waitpid(0,&status,WNOHANG);
	sprintf(exit,"\npid %d exited.\n",pid);
	if(WIFEXITED(status))
	{
		int ab=WEXITSTATUS(status);
		if(ab==0)
			sprintf(exit_status, "normally\n");
		else
			sprintf(exit_status, "abnormally\n");
	}
	if(pid>0)
	{
		write(2,exit,strlen(exit));
		write(2,exit_status,strlen(exit_status));
		commandprompt(actualpath);
	}
	free(exit);
	return;
}
#include "oth.c"
int main(int argc, char** argv)
{
	signal(SIGCHLD,handler);
	char str101[1024],str1[1024];
	while(1)
	{
		commandprompt(argv[0]);
		scanf(" %[^\n]%*c",str101);
		hist(str101,-1,0);
		char* token = strtok(str101,";");
		while (token != NULL)
		{
			for (int i = 0; i <strlen(token); ++i)
			{
				str1[i]=token[i];
				if(i==strlen(token)-1)
					str1[i+1]='\0';
			}
			long long int l=strlen(str1),ptr=0,flag_bg=0;
			while(str1[ptr]==' ')
			{
				ptr++;
			}
			if(str1[ptr]=='c' && str1[ptr+1]=='d')
			{
				ptr+=2;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					while(str1[ptr]==' ')
					{
						ptr++;
					}
					cd(str1,ptr);
				}
				else
				{
					printf("%s: command not found\n",str1);
				}
			}
			else if(str1[ptr]=='p' && str1[ptr+1]=='w' && str1[ptr+2]=='d')
			{
				ptr+=3;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					f_pwd(str1,ptr);
				}
				else
				{
					printf("%s: command not found\n",str1);
				}
			}
			else if(str1[ptr]=='e' && str1[ptr+1]=='c' && str1[ptr+2]=='h' && str1[ptr+3]=='o')
			{
				ptr+=4;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					while(str1[ptr]==' ')
					{
						ptr++;
					}
					f_echo(str1,ptr,l);
				}
				else
				{
					printf("%s: command not found\n",str1);	
				}
			}
			else if(str1[ptr]=='l' && str1[ptr+1]=='s')
			{
				ptr+=2;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					while(str1[ptr]==' ')
					{
						ptr++;
					}
					list(str1,ptr);
				}
				else
				{
					printf("%s: command not found\n",str1);	
				}

			}
			else if(str1[ptr]=='p' && str1[ptr+1]=='i' && str1[ptr+2]=='n' && str1[ptr+3]=='f' && str1[ptr+4]=='o')
			{
				ptr+=5;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					while(str1[ptr]==' ')
					{
						ptr++;
					}
					pinfo(str1,ptr);
				}
				else
				{
					printf("%s: command not found\n",str1);
				}
			}
			else if(str1[ptr]=='h' && str1[ptr+1]=='i' && str1[ptr+2]=='s' && str1[ptr+3]=='t' && str1[ptr+4]=='o' && str1[ptr+5]=='r' && str1[ptr+6]=='y')
			{
				ptr+=7;
				if(str1[ptr]==' ' || str1[ptr]=='\0')
				{
					while(str1[ptr]==' ')
					{
						ptr++;
					}
					hist(str1,ptr,1);
				}
				else
				{
					printf("%s: command not found\n",str1);
				}
			}
			else if(str1[ptr]=='e' && str1[ptr+1]=='x' && str1[ptr+2]=='i' && str1[ptr+3]=='t' && (str1[ptr+4]==' ' || str1[ptr+4]=='\0'))
			{
				ptr+4;
				exit(0);
			}
			else
			{	
				othfunc(str1,ptr);
			}
			token = strtok(NULL,";");
		}
	}
}
